### DISCUSSION TITLE
Allow zone to be a polygon

### DISCUSSION THREAD

**Discussion by Misiu** [home-assistant/architecture#1014] [Entity Models]
## Context
Today, we can create Zones that are based on center point location and radius, this makes it hard to create Zones for places that are near to each other, like buildings, or parts of your backyard, because sometimes the circles may overlap each other:
![image](https://github.com/home-assistant/architecture/assets/1741838/a0739b6d-dc97-423c-9125-b22d38a055a5)

## Decision
We introduce two new properties to the Zone entity model:
 - `zone_type` which can be `circle` or `polygon`, with `circle` as the default to be backward compatible
 - `points` which takes a list of points that define a polygon

## Consequences
Give the end users the ability to specify non-circular zones.
The change will allow picking zone type when creating new zone, or when editing zone.

## Motivation:
With this change, we will be able to create such zones:
![image](https://github.com/home-assistant/architecture/assets/1741838/12d7ff8a-db82-4621-8cd7-9cbd2efe3ac0)

## Additional informations
After getting approval, the next step would be to finish the implementation and move to frontend.
Ideally when adding or editing zones on the frontend the end user should have two options (circle, polygon). When choosing a circle he will get the functionality he has now. When choosing a polygon he should be able to create points that will form polygon for the zone (like in this demo: https://www.geoman.io/demo)
Of course, the zone type can be changed so users can move from circle zones to polygon zones. The editor should also allow editing zones (adding/moving/deleting points).

The original solution proposed using https://www.npmjs.com/package/@geoman-io/leaflet-geoman-free, but maybe something similar can be used.

I've created a PR (https://github.com/home-assistant/core/pull/105729) that uses a simple ray-casting algorithm as suggested here: https://github.com/home-assistant/architecture/discussions/954#discussioncomment-7768724

This discussion is similar to https://github.com/home-assistant/architecture/discussions/954, so a common way of storing and processing coordinates would be useful for both solutions.

**Comment by dbolger** [comment]
+1, This feature would be particularly nice for me. When I walk around our lake loop where I live, my phone just barely catches the zone I call 'Home' and causes automations to fire off. The lake loop isn't a perfect circle, so I'd need to define two zones to cover the area I walk to prevent the trigger, or I'd need to make my zone for my home smaller, which is less than ideal since doing so would make automations that fire off when I actually do arrive home happen later than usual.

I also park my car in different places occasionally, sometimes inside the 'Home' zone and sometimes right outside of it (despite it being here. I could make more circles, but even that isn't as granular as I'd like.

**Comment by Krispkiwi** [comment]
+1 i need this

**Comment by OdinVex** [comment]
+1, my properly is oddly shaped and I'd even like to specify a zone within another (a "lot" versus an owned property such as an RV's allotted space in a much larger area).

**Reply by Misiu** [reply]
I think that right now zone in zone is implemented and it's working. But it must be a circular zone :)
Geometry calculation isn't that hard when it comes to polygons, but it must be fast and that's the challenge.

**Reply by OdinVex** [reply]
I think any pre-made 'point in polygon ?' library would do.. Preferably I'd like to even see some height and use polygon volumes rather than areas. 'on the third floor'...etc.

**Reply by Baael** [reply]
I am using segment tree like approach in js, basically making boundary box - getting longest sides of polygon on both axes, and doing computations only when object is within boundary. It is fast enough to handle all points from cloud points of 6 radars (IWR6843) realtime and in nodered.
But shapely is using GEOS and should be quite fast (https://libgeos.org/)

**Reply by Misiu** [reply]
@Baael according to https://github.com/home-assistant/architecture/discussions/954#discussioncomment-7768827 Shapely doesn't work on armhf based systems , not sure how to test this.

**Reply by GottZ** [reply]
> but it must be fast and that's the challenge.

what is the necessity for speed here?
my home assistant runs on a beefy intel xeon cpu..
doing a simple partitioned poly lookup isn't that challenging in comparison. especially since gps coords don't arrive each second.
don't tell me you run HASS on a raspberry pi 2..
wouldn't it be more reasonable to simply add a disclaimer to high compute demand?

**Comment by Baael** [comment]
Big kudos to @Misiu for his work.

Where do we stand? What needs to be done to move this forward?

In my opinion, this is one of the most needed changes. Circular zones simply won't serve their purpose and can only be treated as general areas, like near a station. When you have two parking lots next to each other and you want to implement two different automations depending on the parking lot - it's impossible.
The same goes for apartments, gardens, and driveways.
I myself recently had two apartments next to each other, and the current implementation forced me to write my own hack.

Not to mention that it's currently impossible to drive near the house and parking lot going from preschool to work, when the road is nearby, without triggering the automation.
In such cases, circular zones are completely useless.
GPS accuracy has significantly improved compared to a few years ago, plus I have installed a precise version myself, with an accuracy of ~5cm in the area of my home.

**IMO even if for some reason the FE (geoman) is problematic to implement in HA in this case, the backend should implement it, giving possibilities to users, and then the FE can be delivered as a custom component from HACS or whatever.**

**Comment by MichelGerding** [comment]
Would love to see this implemented. 

The current system of only allowing circles is inherently flawed. not  a single zone i am tracking is a circle and using multiple circles for a single zone is not a real solution. like mentioned before i want to track multiple zones that are close to each other which becomes close to impossible with circles. the primary locations where i am in need of this is my local race track. i would like to be able to differentiate between the track or the road that is near it, the karting track, or the buildings around it. or my collage campus where i want to track specific buildings or even wings of buildings and parking lots. The way this proposes would be an amazing way to 

Like @Baael says if the FE is the issue then i would still love it it was implemented. The current problem for the frontend is how would it the data be represented. Because if why implement something that doesn't have uses. 

If the problem is in what format to store the data i would like to suggest the [GeoJson](https://datatracker.ietf.org/doc/html/rfc7946) format. its easy to read and debug and supports whatever you want. due to the simplicity of the storage of the data and the large support both the current implementation of circles and the proposed situation of polygons can be rendered. GeoJSON even supports excalves. If geospatial data would be stored and rendered like this it would bring support for a large amount of different opportunities in the map.

For testing in which zone we are the ray-casting algorithm suggested in issue [#954 (comment)](https://github.com/home-assistant/architecture/discussions/954#discussioncomment-7768724 ) you can support en-/exclaves by simply testing all zones/polygons. rudimentary tests on a rpi zero w shows that millions of zones can be queried a second. 

So to reiterate. this would be absolutely amazing if this can be implemented. this would open a large amount of opportunities that the current zones don't offer. As this proposal only extends on the current spec for zones it should not break a single integration or feature. the only points that i could see that would prevent this would be lack of support for the frontend or the want for a better defined common way to store geospatial data.

**Reply by Baael** [reply]
GeoJSON looks great I am for it, I am using also line crossing to determine the direction of movement, but I can imagine a lot of other uses: geofencing, distances, directions, paths. 
+1

**Comment by Menelao147** [comment]
I would really like to see polygonal zones implemented. I have several zones in HA and with the circle I can't define their area correctly

**Comment by mandohunter** [comment]
I'd love to see polygonal zones implemented, as well. I have a large rectangular property (4+ acres) and the "home" zone doesn't cover enough of it without "encroaching" on my neighbors' properties. Loving the GeoJSON as a part of implementing.
+1

**Comment by cohaolain** [comment]
This is a great idea and something I would really appreciate in Home Assistant if it can be gotten over the line - I really hope it gets some attention soon!

Currently, I have a few zones that are close together, and the limitation of circular zones prevents me from defining them accurately. I also have zones that are essentially "strips" which circles are not suitable for at all. 

Being able to define zones as polygons would give much more flexibility and allow for more precise zone design. Geojson would allow easy definition of complex zones where needed.

Thanks for your work on this!

**Comment by DarkFox** [comment]
In the meantime, here's a simple custom template that lets you create your own Geofence template sensors:
https://gist.github.com/DarkFox/5dff2c70edce530caa2a986ab2d6a776

**Reply by MichelGerding** [reply]
As an alternative i have been working on a custom integration providing this functionality. together with a map to edit the zones. they can be found at the following locations.
integration: https://github.com/MichelGerding/Homeassistant-polygonal-zones
map addon: https://github.com/MichelGerding/Homeassistant-polygonal-zones-addon

**Reply by ToroNander** [reply]
Works perfectly! THX!

**Reply by stephenshutters** [reply]
> I have been working on a custom integration providing this functionality.

Unfortunately, @MichelGerding, your integration [is no longer functional](https://github.com/MichelGerding/Homeassistant-polygonal-zones/issues/11).

**Reply by stephenshutters** [reply]
@DarkFox, your links appear to be broken.

**Reply by DarkFox** [reply]
@stephenshutters Yeah, I made my repo private recently.

I've made a gist instead: https://gist.github.com/DarkFox/5dff2c70edce530caa2a986ab2d6a776

**Comment by oerix** [comment]
How is this still not a thing in HA?

**Comment by KevinCathcart** [comment]
One of the big challenges here is that the most common GPS tracker used is mobile phones.

Both IOS and Android only support circular geofences at an OS level. The companion apps set these up from circular zones. But obviously this can only work with circular zones.

Beyond geofencing, the Android companion app mostly uses OS provided background location updates, which are reasonably frequently, and of varying accuracy. These might not always be able to determine zone location as fast or accurately as geofencing can with circular zones, but it might be good enough for some users in conjunction with polygonal zones.

However, on iOS, the OS level background updates ("major location updates") are simply not good enough, so the companion app also defaults to getting and sending location information with other periodic sensor updates, which is by default every 15 minutes. That would be a problem for zones, as most people won't want an up to 15 minute delay detecting a user is in a zone. But the only alternatives are more frequent periodic background updates (causing more battery drain) or using circular geofences, which won't work with polygon zones.

**Reply by OdinVex** [reply]
> One of the big challenges here is that the most common GPS tracker used is mobile phones.
> 
> Both IOS and Android only support circular geofences at an OS level. The companion apps set these up from circular zones. But obviously this can only work with circular zones.
> 
> Beyond geofencing, the Android companion app mostly uses OS provided background location updates, which are reasonably frequently, and of varying accuracy. These might not always be able to determine zone location as fast or accurately as geofencing can with circular zones, but it might be good enough for some users in conjunction with polygonal zones.
> 
> However, on iOS, the OS level background updates ("major location updates") are simply not good enough, so the companion app also defaults to getting and sending location information with other periodic sensor updates, which is by default every 15 minutes. That would be a problem for zones, as most people won't want an up to 15 minute delay detecting a user is in a zone. But the only alternatives are more frequent periodic background updates (causing more battery drain) or using circular geofences, which won't work with polygon zones.

That's an iOS/Android issue, not an HA issue. I use GPS-only (no location provider garbage or other "telemetry" (read spying) means) so I'm used to cold-start GPS. It's what I want anyway. If people want to use circles, let'em. If they want to use polygons...why screw them over?

**Reply by KevinCathcart** [reply]
Its a Home Assistant issue in so far as it is proposing to change core to support zone types which will provide a poor user experience to users using the companion apps as trackers in the default configuration.

Merely allowing non-circular zones would require pretty massive companion app documentation changes, as the documentation would need to be clear that such zones don't count for the zone based updates (which are geofence powered), and for most users would prove to be meaningfully slower to update and less accurate than using geofence capable (circular) zones. Sure that perhaps isn't true of users who go out of their way to disable non-gps location services, and thus will either get infrequent location updates or experience elevated battery drain, but that is also an uncommon configuration.

The mere fact that the documentation would need to highlight a pretty significant caveat to using these makes it a hard sell. Add on top of that the support headaches from users who have not read the documentation and don't understand why these polygon zones seem slower and less reliable than the circular ones, and it becomes an even harder sell. 

Then there is the big question of if this is only supported in core, or if the companion apps will also support interpreting polygon zones, so that they can report back just the zone name rather than full GPS coordinates (a feature already offered for circular zones). Letting the companion apps interpret polygonal zones and report them back would be better for privacy, but means that three copies of polygon zone evaluation (in three different languages) would need to be written and maintained. Having just core support them only needs one copy, but adds additional edge cases with respect to things like overlapping circle and polygon zones when zone name reporting is used.

**Reply by OdinVex** [reply]
> ...

Not sure why anyone assumed a circle to begin with considering it's trivial to conclude whether an xy-point is within a polygon. It's a rather basic function in video games, even.

> uncommon configuration

User's choice (especially from a privacy-considered point of view). More Android ROMs are giving users the option to use alternative location providers or even disable it entirely and simply use GPS like a normal, unnumbered-for-profit person. Agreed it's uncommon as most people aren't aware of the effects of Google's information trove and American Government/Military interests in it (and I say this as someone living in NY, USA).

As for documentation: "Documentation update: Zones can now be polygons or circles, polygons may be slower by a few cycles of processor instruction but imperceptible to anyone that doesn't mind a nano-second delay." Pretty straight-forward. Client literally just needs the polygon shape (includes size), location, and GPS coordinate (however they prefer, via GPS or "Google"). Use the same math used in calculating "does xy point exist within this polygon" on an xy graph. It's easily extensible to xyz if you consider elevation too (high-rise/apartment buildings anyone?) but that's a bit pedantic at that point (in my opinion, but if someone wants it, I'm no one to knock it).

It can be calculated on any end as the data transmitted and the polygon location is easily made available to either.

"User is in zone A ("Home", subset zone of B) and B ("Apartment Building") and D ("Rear of Apartment Building") intersections (alias: the community pool)."

"User is in zone F ("Family Member's Home", subset zone of B) and B ("Apartment Building") intersection (alias: Uncle Whoever's Home)."

Zone identification can be sets with optional aliases even, clients can report back GPS coordinates (and receive polygon info, can calculate locally) or in cases where clients are unaware of GPS info but core has some way of tracking them (sensor not known to client, such as parental supervision of clients via an external means) core can calculate and inform client.

I really don't see it as a complicated thing but I'm also not the one developing for HA or it'd include the `ath12k` driver already so I could use WiFi 7 adapters with HA and probably make the firmware folder a mount point that simply copies firmware from updates to it (if missing partition) so users could easily add their own and override or downgrade. Is what it is. :/

...But I think the bottom line is that it's desired. I am pretty sure, quite willing to bet even, that most zones aren't going to be circular and it's desired. Low-priority, perhaps, but creating 15 different circles and writing scripts to "virtually merge them together" to try to get a close-enough polygon "zone" for things to match with just reaaally sucks. Maybe I'm just expecting too much from HA. (On the polygon zone things, not the firmware, that's a big issue.)

**Reply by KevinCathcart** [reply]
When I'm talking slower I mean that that on iOS with the default settings the a polygon zone will likely need to wait for next 15 minute sensor refresh before the the app does a background location refresh and gets new coordinates which will let it detect you have entered/left the zone. Versus usually under a minute to be notified of entering a circular geofence. Up to 14 minutes slower (and 6 minutes on average) to be recognized as entering a zone is likely to be unacceptable to many people. If users were to try to make it almost as fast by setting the periodic sensor frequency to every minute, the extra battery drain likely becomes rather noticeable, since it would be using up to 15 times as much power. Could this be fixed by having the app poll location more often, but just not always report it unless a zone change happens, or it is time for sending periodic sensors? Maybe, but that just means this would need even more work.

The difference would likely be a smaller on Android with default settings (app and OS), since it typically gets updates every 3 minutes or less, although details vary by location provider. 

My point here basically comes down to this being more a lot more obnoxious than it naively sounds to deal with in the mobile companion apps. Getting all of this coded and tested would be more work than it sounds. Maintenance and support burden is also likely significantly greater than you expect. All of which means it is not an especially appetizing project for the devs to take on. Beyond just the core and mobile app work, they would also need to create a GUI polygon zone editor (which is the sort of thing that sounds easy, but can end up taking 6 person-months to get solid), figure out how to best display it, etc. Sure it would be a nice improvement from the end users perspective, but I'm skeptical if the core devs will find it worth the hassle in the foreseeable future, especially vs other high impact changes they could be making.

**Comment by OdinVex** [comment]
> iOS/Android...

Still a "them" problem. Besides, there's an easy work-around. When a polygon is communicated (first time) it can be "low-resolutioned" into just a circle for them and left alone for high-poly users. Let users poll as often as they want (every second, who cares, it's their device). So regular solid-lined polygons for "high-resolution" users and a dash-lined circle (overlayed on the editor?) for "low-resolution" users.

I'm highly skeptical the core developers would accept it simply because I'm interested in it. >_>...I'm actually skeptical about their entire push to let Google and other Big Tech/corp into our private networks with their "upsell to cloud, moniez$$$" updates and nagware you can't remove unless you adblock/compile it yourself. Glad to have HACS. It's also a pain to have to rewrite packets to prevent so much call-home by HA.

(Edit: My point is there's obviously a difference in priority for...*them* versus the ones here who've posted in favor of a different feature.)

**Comment by SaturnusDJ** [comment]
Can we, as a first step of improving zones, allow multiple circles to be assigned to a single zone?
```
- name: Work
  icon: mdi:briefcase
  - location
    latitude: 1.2345
    longitude: 6.7890
    radius: 20
  - location
    latitude: 2.2345
    longitude: 7.7890
    radius: 20
```

On the device hardware to software level no changes are needed, just Home Assistant has to merge them in the entity.
The idea is not to then use 29 circles to create a near perfect square but to get better coverage and prevent unwanted check-ins.

(Random example data)
Now:
<img width="900" height="600" alt="OSM1" src="https://github.com/user-attachments/assets/e88e8e41-26ad-4c37-b30f-1ab6068542fd" />

Multiple circles:
<img width="900" height="600" alt="OSM2" src="https://github.com/user-attachments/assets/5ac49c83-cc5f-4abc-8d3b-0483094eecd3" />

**Reply by stephenshutters** [reply]
This would be helpful!

**Comment by bgoncal** [comment]
Year ago when I tried to achieve polygon zones in a different iOS project it was painful, several circular zones neede to be observed to achieve the final result, maybe there is something more modern nowadays but I couldn't quickly find, if anyone knows or find something, let me know.

**Reply by OdinVex** [reply]
> Year ago when I tried to achieve polygon zones in a different iOS project it was painful, several circular zones neede to be observed to achieve the final result, maybe there is something more modern nowadays but I couldn't quickly find, if anyone knows or find something, let me know.

It's a [point in polygon](https://wikipedia.org/wiki/Point_in_polygon#Point_in_polygon_queries) issue.

**Reply by bgoncal** [reply]
Sorry, what do you mean? that is possible? iOS has methods to subscribe for enter/exit, and those are for circular zones only as far as I know

**Reply by OdinVex** [reply]
> Sorry, what do you mean? that is possible? iOS has methods to subscribe for enter/exit, and those are for circular zones only as far as I know

No clue about iOS, but it's how anyone would judge whether a point is in a polygon. Edit: Sounds like the primitive nature of Android/iOS is a problem. Maybe an circle can be generated for polygons just for their poor capabilities (I know it's about Android/iOS hiding how the battery usage of their "telemetry"/spying, frankly such behavior shouldn't be wall-gardened in favor of their own software such as for selling location).

**Reply by OdinVex** [reply]
In the interim to an alternative operating system for phones free of evil capitalists, maybe we could roll polygons and behind the scenes provide a "mobile-version" polygon (or "converted to a group of circles as a single entity"?)

**Comment by emontnemery** [comment]
We discussed this in a core meeting last week. The limitation in the iOS and Android APIs where they only support circular zones means we can't accept the proposal at this time since it would break the native apps waking up when entering or leaving a zone.

**Reply by OdinVex** [reply]
> We discussed this in a core meeting last week. The limitation in the iOS and Android APIs where they only support circular zones means we can't accept the proposal at this time since it would break the native apps waking up when entering or leaving a zone.

If more than one zone can be supported at the same time then a polygonal zone could be divided into multiple smaller zones (slightly less accurate but functional at least) with some overlap even? Treated logically as a single contiguous zone?

**Reply by Misiu** [reply]
That makes sense given the current platform limitations.

Before closing this direction entirely, would it be reasonable to get feedback from the mobile maintainers on an alternative approach?

Specifically: using native circular geofences on iOS and Android purely as wake-up triggers (for example, derived from a polygon’s bounding circle), while treating the polygon itself as the authoritative geometry and validating zone membership client-side after the app wakes up.

This would preserve the existing OS-level behavior (apps still wake on enter/exit of circular regions) while avoiding a hard dependency on system-level polygon support.

**Reply by emontnemery** [reply]
@Misiu What would the app do when you're inside the bounding circle, but not yet inside the zone as defined by a polygon, would it keep periodically waking up? I think any such approach would either ruin the battery life or poll too seldom to be useful.

But let's hear from @bgoncal and the other app developers.

**Reply by Misiu** [reply]
@emontnemery this was just an initial suggestion, not sure if there are better ways to handle it. Waiting for Android and IOS API isn't a solution and probably they won't add support for polygons.
I've tried a couple of approaches in Core, but I wasn't aware that this is used in mobile apps as well.

**Reply by OdinVex** [reply]
> What would the app do when you're inside the bounding circle, but not yet inside the zone as defined by a polygon, would it keep periodically waking up? I think any such approach would either ruin the battery life or poll too seldom to be useful.

Wouldn't it be like any other circle? You enter a circle, and if it's part of a polygon the circle is considered the polygon, otherwise it's just a regular circle. If you leave one of the "subcircles" of a polygon and enter another just see if it's part of the polygon and do nothing. Just warn users of the slight extra drain from polygonal zones. My phone lasts about two weeks on a charge with high-frequency reporting of HA sensors, so maybe I care less about drain than others might, but I don't imagine it being any more than someone who has a few zones all close together entering and leaving them.

Edit: I envision polygonal zones being calculated as lumpy cloud shapes that are just overlapping circles). Slight accuracy loss around some areas of perimeter, but still very useful where I live.

**Reply by bgoncal** [reply]
@emontnemery The app cannot use the polygon as the source of truth, it can keep using a circle if provided but it wont take into the consideration the polygon after entering the circle.
Implementing a custom polygon "check" whenever the user enters a circle would mean, as you previously said, battery waste and possible backround interruptions depending how long the operation takes.
I am "against going against" the platform supported methods, so for now since iOS only supports circle, thats what I would vouch for.

**Reply by OdinVex** [reply]
Would the drain be that much if it is first simply check as to whether it belongs to a polygonal type or not? "if circle is a zone then do regular zone stuff, else a polygonal circle (determine which polygonal zone, trigger for that zone)". Seems lightweight to me.

**Reply by bgoncal** [reply]
Depends how often the person moves away and back, but there are also other limitations, we share background activity and opportunities between other features as well such as to report other sensors , if location takes a bunch time from this "opportunity window" we can ruin the output completely and not be able neither to determine if its inside the polygin either to report the other sensors. But regardless of that, the complexity would be high to maintain, specially because depending on iOS version I am pretty sure we would need to adapt the approach based on APIs available for that OS version.

**Reply by OdinVex** [reply]
Maybe it could be implemented as experimental, drain-acceptance feasibility study and such? Edit: There's already an option for frequent sensor reporting, so maybe some wouldn't mind it.

**Reply by bgoncal** [reply]
If someone has time to play with it, I would be ok having as an experimental feature and fully opt-in.

**Reply by Baael** [reply]
I think the discussion is currently conflating three separate concerns:

1. **Zone geometry / data model**
2. **How a specific client detects presence**
3. **Backward compatibility with the current mobile app behavior**

The fact that the iOS and Android companion apps currently rely on native **circular** geofences should not, by itself, limit the geometry model of the platform as a whole.

Mobile apps are only one class of client. They are not “the device model” for all location-aware use cases. A platform-level zone model may also be relevant to:

* vehicles
* mowers
* vacuums
* GPS trackers / wearables
* server-side presence evaluation
* future integrations that do not depend on mobile OS geofencing at all

Because of that, I do not think polygon support should be rejected purely because the current mobile apps only support circles natively.

If backward compatibility is the concern, then the solution does not have to be “replace circles” or “change existing zones semantics”. A cleaner approach would be to introduce a **new layer / new type**, e.g. `polygon_zone`, exactly as originally proposed.

That would allow:

* existing circular zones to remain unchanged
* current mobile apps to continue supporting only circular zones
* polygon zones to be supported by integrations, devices, and server-side logic that can use them
* capability differences to be explicit instead of forcing the whole platform to match the lowest common denominator

In other words: lack of support in one client should not define the global data model.

If the real concern is maintenance burden, UX clarity, or event semantics, I think that should be stated explicitly. Those are valid product concerns. But they are different from saying polygon zones are not viable because mobile apps only support circular geofences.

A possible path forward could be:

* keep current `zone` behavior unchanged for circular/native mobile compatibility
* add a separate `polygon_zone` type/layer
* let clients/integrations declare what they support
* evaluate polygon membership where it makes sense (server-side or device-side), without requiring the companion apps to emulate unsupported native behavior

That would preserve backward compatibility while still allowing the platform to support richer geometry where available.

This way, existing mobile behavior remains untouched, the broader platform is not artificially limited by current app constraints, and if native polygon support ever arrives on mobile, the architecture is already ready for it.

**Reply by mbrevda** [reply]
> We discussed this in a core meeting last week. The limitation in the iOS and Android APIs where they only support circular zones means we can't accept the proposal at this time since it would break the native apps waking up when entering or leaving a zone.

There seems to be a much broader use case beyond mobile-based positioning. There are plenty of other sources of data which can provide polygons, and the lack of mobile positioning shouldn't block, say, the map card from supporting these.

**Reply by crua9** [reply]
@emontnemery 

You can solve this easily by doing several things. 


1. The "Wake Up" (Outer Circle)

Just have the wake up circle as an invisible layer outside of the geofence. Like basically take the geo fence the user draws, then make it bigger to cover whatever range. You could have it in the users hands what that range should be, but also a default. But have this wake up area invisible. Like it doesn't have to be a "circle". Just expand what is there by a little. Specifically, make a **Minimum Enclosing Circle (MEC)** with a user-defined offset.

This buffer zone is what is pushed to the mobile devices. The user shouldn't be able to see this zone to prevent confusion. Or maybe during setup there might be a way but after not? Or maybe a toggle to let the user see it so they can test? Maybe have it as a dash line or something else? IDK this is a UI thing. 

- **The Phone's Role:** You define a large, "dumb" circular geofence (e.g., 100 meters). 
- **The Action:** The OS passively monitors this. When you cross it, the phone "wakes up" the Home Assistant companion app. 
- **The Result:** The app starts sending high-accuracy GPS data to the server only after you enter this "Warm-up Zone."

2. The "Precision Check" (Inner Polygon)

Upon crossing the "Wake Up" boundary, the app triggers a Background Task Extension (iOS) or Foreground Service (Android). The app begins a 25-second "sprint" of high-accuracy GPS updates to the HA server.

- **The Server's Role:** Home Assistant receives the high-accuracy stream. It uses a "Point-in-Polygon" check to see if you are actually inside the oddly shaped property line.

The server evaluates the points. If the user enters the polygon, the state changes to home. If the 30-second timer expires before the polygon is reached, the app gracefully returns to "Significant Location Change" monitoring.

If the user stays within the wake area, but not going in the main area. Then the logic should allow for re-triggering. If the app goes back to sleep but the OS sees the user is still within a specific "proximity" or moving toward the center, it should be able to trigger another 30-second window after a short cooldown. During those 30 seconds of execution time, the app tells iOS: "Stop watching the 500m circle. Instead, start watching this 100m 'Proximity' circle." The app should only swap the geofence if the Speed and Heading suggest the user is moving toward the center. If they are just "grazing" the edge, it stays as the large circle. This prevents an issue if the user turns around. This also deals with the 20-region limit. The 'Wake Up' circle should not be a secondary entity. Instead, the integration should utilize Dynamic Region Re-registration.

If you're on the edge then go to 3.

3. Solving the "Bouncing" (Hysteresis)

To prevent it from flickering as you walk along the edge of the fence, you use Hysteresis:

- **Entry Rule:** You aren't "in the zone" until you are 3 meters inside the polygon.
- **Exit Rule:** You aren't "Away" until you are 10 meters outside the polygon.
- **The "Speed" Factor:** If the server sees you are moving at 40 mph, it can "predict" you are passing through and ignore a 1-second blip inside the zone to prevent a false trigger.

I would suggest letting the users edit these ranges, but there being some default. But basically this makes it where the user when they go inside the geo fence area at a low speed. They will only be "in" the geo fence if they are beyond a point. But to prevent bouncing it takes them a little bit more to leave it. 
This is why it is important to give the user the ability to pick the range. Like it might be more important for it to trigger on than off, or off than on. 


From my understanding, Apple is much stricter than Android. If an app tries to run high-accuracy GPS in the background for more than a few minutes, iOS will often kill the process entirely, even if you are inside a "Wake" zone. To solve for this Instead of the Home Assistant app staying "awake" to watch your GPS, it tells iOS: "Hey, I'm going to sleep now. Here is a circular coordinates and a radius. If the user crosses this line, please wake me up for 10 seconds."

- The OS handles the heavy lifting. Since the OS is already monitoring cell towers and Wi-Fi for its own needs, this uses almost zero extra battery.

Apple allows apps to periodically fetch data in the background. Home Assistant uses this to update your sensors (battery level, steps, etc.). If you are inside your "Outer Circle" (the Wake Zone), the app can try to request a "Background Task Extension." This gives the app an extra 30 seconds of high-accuracy GPS time. If the app hasn't determined you are in the "Inner Polygon" by then, it has to shut down or risk being "blacklisted" by the OS for future background tasks.


By using the Background Task Extension and the Watchdog Timer, the app stays within Apple's strict 30-second execution window. It avoids being "blacklisted" because it explicitly tells the OS when it is done with the high-accuracy task.



TLDR:

- **The "Wake Up" Buffer:** An invisible, large circular geofence—calculated as a Minimum Enclosing Circle (MEC)—that tells the phone to "wake up" the app as you approach.
- **The "Precision" Sprint:** Once the phone enters the buffer, it triggers a 30-second high-accuracy GPS burst. The server (Home Assistant) then performs a Point-in-Polygon check to see if you are actually inside the irregular property lines.
- **Anti-Flicker (Hysteresis):** To prevent "bouncing" between Home/Away, you use offset rules (e.g., 3m inside to trigger "Home," 10m outside to trigger "Away") and speed/heading logic to ignore false blips.
- **OS Efficiency:** It bypasses battery drain and Apple's strict background limits by using native OS monitoring to handle the heavy lifting while the app stays "asleep" until the last moment.

**Reply by gitmayer** [reply]
I still can't fathom why the mobile app is causing this not to happen. As many others have pointed out, some of us just have lat and lon data being reported back and not using zones within the phones. I want my HA server to determine based on polygons like it currently does with center and radius. This should not be so hard or political.

**Reply by OdinVex** [reply]
Bingo. Server can then send the results of calculation back to phone...so simple, but ugh..

**Comment by GottZ** [comment]


**Reply by OdinVex** [reply]
>_> It's been thought of, repeatedly in this Issue as anyone that's read it can see.

**Comment by jvmahon** [comment]
The Matter Application Cluster Specification Version 1.5 has added a new Zone Management Cluster which allows user-defined 2-D Cartesian zones with assigned use types (Motion detection, Privacy, and Focus). 

It seems that Home Assistant's Matter support will, at some point, need to support this part of the Matter standard. So, as you are considering the use of polygon zones, it seems advantageous to start with the Matter model and use that (or at least make sure that is supported by and encompassed in whatever might get added to Home Assistant.).

**Reply by ximex** [reply]
from the matter spec 2.14.4.1. TwoDimensionalCartesianZone Feature:
> The origin (0,0) SHALL be located at the top left of the Cartesian plane, with positive X and Y values moving right and down across the Cartesian plane respectively.

2.14.5.6. TwoDCartesianVertexStruct Type defines unsigned int for the coordinates

this wouldn't fit to coordinates for the world (WGS84/EPSG:4326)

**Comment by ukanuk** [comment]
It would be awesome for Home Assistant to add isochrones from a location, which would only be possible with polygonal zone support.

Now that Home Assistant is using OpenStreetMap for its mapping provider, it seems it could be a little easier -- openrouteservice.org can already do the calculations using OpenStreetMap data, for example [15min by car from Koln: ](https://maps.openrouteservice.org/#/reach/K%C3%B6ln,NW,Germany/data/55,130,32,198,15,97,4,224,38,9,96,59,2,24,5,192,166,6,113,0,184,64,118,0,232,0,102,32,38,1,56,40,3,128,26,1,89,140,34,179,168,25,128,22,14,69,164,8,0,117,79,2,34,108,57,65,135,72,131,52,92,160,0,217,165,193,222,161,0,140,0,216,56,111,193,183,188,196,1,205,112,5,160,166,196,189,53,109,139,224,11,235,192,23,148,0,182,184,213,209,8,160,39,186,89,121,228,32,193,145,229,225,81,188,121,192,32,1,93,165,160,34,240,1,85,16,195,209,96,0,8,1,149,80,208,177,35,249,160,32,0,205,225,229,209,112,65,97,161,224,0,220,145,12,77,131,101,121,161,144,141,208,1,245,195,249,202,240,132,93,202,90,219,13,122,1,180,41,72,1,117,120,144,100,106,67,112,39,136,236,236,128)

<img width="978" height="686" alt="image" src="https://github.com/user-attachments/assets/94d22623-6362-4580-a8be-b444a1beb6ec" />
